File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int bfs(const vector<string>& maps, pair<int,int> start, char target) {
    int n = (int)maps.size();
    int m = (int)maps[0].size();
    vector<vector<int>> dist(n, vector<int>(m, -1));
    queue<pair<int,int>> q;

    dist[start.first][start.second] = 0;
    q.push(start);

    int dx[4] = {1, -1, 0, 0};
    int dy[4] = {0, 0, 1, -1};

    while (!q.empty()) {
        auto [x, y] = q.front(); q.pop();

        if (maps[x][y] == target) return dist[x][y];

        for (int dir = 0; dir < 4; dir++) {
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
            if (maps[nx][ny] == 'X') continue;         // 벽
            if (dist[nx][ny] != -1) continue;          // 방문됨

            dist[nx][ny] = dist[x][y] + 1;
            q.push({nx, ny});
        }
    }
    return -1; // target 도달 불가
}

int solution(vector<string> maps) {
    int n = (int)maps.size();
    int m = (int)maps[0].size();

    pair<int,int> S{-1,-1}, L{-1,-1};

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (maps[i][j] == 'S') S = {i, j};
            else if (maps[i][j] == 'L') L = {i, j};
        }
    }

    int d1 = bfs(maps, S, 'L');   // S -> L
    if (d1 == -1) return -1;

    int d2 = bfs(maps, L, 'E');   // L -> E
    if (d2 == -1) return -1;

    return d1 + d2;
}

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;
class Solution {
    int[] dx = {-1, 1, 0, 0};
    int[] dy = {0, 0, -1, 1};
    
    public int solution(String[] maps) {        
        int[] s = new int[2];
        int[] l = new int[2];
        int[] e = new int[2];
        
        for (int i = 0; i < maps.length; i++) {
            for (int j = 0; j < maps[i].length(); j++) {
                if (maps[i].charAt(j) == 'S') {
                    s = new int[]{i, j};
                }
                if (maps[i].charAt(j) == 'L') {
                    l = new int[]{i, j};
                }
                if (maps[i].charAt(j) == 'E') {
                    e = new int[]{i, j};
                }
            }
        }
        
        int a = bfs(maps, s, l);
        if (a == -1) return -1;
        int b = bfs(maps, l, e);
        if (b == -1) return -1;
        return a + b;
    }
    
    private int bfs(String[] maps, int[] start, int[] end) {
        int n = maps.length;
        int m = maps[0].length();
        
        boolean[][] visited = new boolean[n][m];
        Queue<int[]> q = new LinkedList<>();
        
        q.offer(new int[]{start[0], start[1], 0});
        visited[start[0]][start[1]] = true;
        
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            
            int x = cur[0];
            int y = cur[1];
            int dist = cur[2];
            
            if (x == end[0] && y == end[1]) {
                return dist;
            }
            
            for (int d = 0; d < 4; d++) {
                int nx = x + dx[d];
                int ny = y + dy[d];
                
                if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
                if (visited[nx][ny]) continue;
                if (maps[nx].charAt(ny) == 'X') continue;
                
                visited[nx][ny] = true;
                q.offer(new int[]{nx, ny, dist + 1});
            }
        }
        
        return -1;
    }
}